Microshift: add disk resize functionality

praveenkumar opened the pull request:
The way this PR works is in following steps
- It grows the disk size usign growpart command
- Resize the physical volume
- Make sure that volume group where `/sysroot` mounted and free space which is going to used by LVMS for persistant volume (pv) creation use equal disk size during the resizing because after `pvresize` all the space goes as free space.

praveenkumar commented:
/retest

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,41 +112,104 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
-		if exitErr.ExitStatus() != 1 {
-			return err
-		}
+	_, _, err := sshRunner.RunPrivileged(fmt.Sprintf("Growing %s partition", rootPart), "/usr/bin/growpart", rootPart[:len("/dev/.da")], rootPart[len(rootPart)-1:])
```
Can you split this in 2 commits, first the introduction of `getRootPartition` and `runGrowpart`, and then the addition of microshift disk resizing? The refactoring introduces a minor regression which is very hard to catch as the diff is hard to read here.
I've done this in https://github.com/cfergeau/crc/commits/issue_3624

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,41 +112,104 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
 
-		logging.Debugf("No free space after %s, nothing to do", rootPart)
-		return nil
+	logging.Debugf("No free space after %s, nothing to do", rootPart)
```
This was only printed when `sshRunner.RunPrivileged("growpart")` returned an error, and if the exit code was 1. After your changes it's also printed when `err` is nil.

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,41 +112,104 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+}
+
+func growLVForMicroshift(sshRunner *crcssh.Runner) error {
+	rootPart, err := getrootPartition(sshRunner, "/dev/disk/by-id/lvm-pv-*")
```
I'd have a strong preference for, for example, getting the device to use using `sudo pvs -S vg_name=rhel -o pv_name --noheadings` or some similar command.
Alternatively, you can change https://github.com/openshift/microshift/blob/bac892b5586edff5687b399475f8ad0e8ec01994/scripts/image-builder/config/kickstart.ks.template#L28 to set a well-known label (see [the --label option to part](https://docs.fedoraproject.org/en-US/fedora/f36/install-guide/appendixes/Kickstart_Syntax_Reference/#sect-kickstart-commands-part)). Then you can use this label to `getRootPartition`.

Passing a wildcard to `getRootPartition` only works because we get the root partition through SSH, which I'd rather not leak outside of `getRootPartition`. And there's nothing in the code erroring out if we are unlucky and there are multiple matches for lvm-pv-*.

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,41 +112,104 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+
+	// Get the size of volume group
+	// $ sudo vgs --noheadings --nosuffix --units g -o vg_size
+	//   39.02
```
I'd remove these last 2 lines of comments.

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,41 +112,104 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+	vgFree, err := strconv.ParseFloat(strings.TrimSpace(freeVG), 64)
+	if err != nil {
+		return err
+	}
```
Please add a helper, `func getVGSize(kind string) (float, error)` which will be called with `vg_size` and `vg_free`.
Might be better to work in bytes directly?

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,41 +112,104 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+		return err
+	}
+
+	// We want to expand the `/dev/rhel/root` vg in case the free size of volume group is greater than half of total size
```
My understanding is that `/dev/rhel` refers to a VG, while `/dev/rhel/root` is a LV. I don't know which of the 2 you want to refer to when you say `the '/dev/rhel/root' VG`

praveenkumar replied:
Ah yes it is LV which need to be extended.

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,41 +112,104 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+
+	// We want to expand the `/dev/rhel/root` vg in case the free size of volume group is greater than half of total size
+	// since that vg contain the `/sysroot` so we need to extend it to accommodate more container images. Free size is
+	// used by logical volume manager storage (LVMS) Container Storage Interface (CSI) provider for storing PVs.
```
PVs means (k8s) Persistent Volume in this comment I think? While in the rest of the function it is a LVM Physical Volume.
Maybe rewrite the last sentence as `The LVM Storage Operator will use the free space for persistent local storage` or something like this. I used https://docs.openshift.com/container-platform/4.14/storage/persistent_storage/persistent_storage_local/persistent-storage-using-lvms.html for inspiration.

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,41 +112,104 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+	// We want to expand the `/dev/rhel/root` vg in case the free size of volume group is greater than half of total size
+	// since that vg contain the `/sysroot` so we need to extend it to accommodate more container images. Free size is
+	// used by logical volume manager storage (LVMS) Container Storage Interface (CSI) provider for storing PVs.
+	sizeDiff := vgFree - vgSize/2
```
Having a hard time to wrap my head around this, might be easier to read as (which should be approximately correct?)
```
  if vgFree > vgSize/2 {
    // run lvextend --resizefs --size $vgSize/2
  }
```

cfergeau reviewed (commented):
Some comments, but overall logic should be fine.
Not fully sure about the UI though, what the size the user specified should refer to? The size of the partitions where they store their images/... ? Or the size of theoverall disk? 
The former is only resized by half of what the user specified if I understood correctly.

praveenkumar commented:
> Not fully sure about the UI though, what the size the user specified should refer to? The size of the partitions where they store their images/... ? Or the size of theoverall disk?
The former is only resized by half of what the user specified if I understood correctly.

@cfergeau it should be the overall disk same like for openshift preset, whatever user provide, half will be added to `/dev/rhel/root` logical volume (LV) which have `/sysroot` mounted and store images and another half would be part of free Volume group (VG) which is used by persistent volume of kubernetes side.

cfergeau commented:
> > Not fully sure about the UI though, what the size the user specified should refer to? The size of the partitions where they store their images/... ? Or the size of theoverall disk?
> > The former is only resized by half of what the user specified if I understood correctly.
> 
> @cfergeau it should be the overall disk same like for openshift preset, whatever user provide, half will be added to `/dev/rhel/root` logical volume (LV) which have `/sysroot` mounted and store images and another half would be part of free Volume group (VG) which is used by persistent volume of kubernetes side.

What I'm asking is if it makes sense from a user perspective. Setting disk size, ssh'ing to the VM, checking df -h will only show half of that size. This is unexpected and different from the openshift preset. Maybe we should change the LV to the size the user asked for?

praveenkumar commented:
>  Maybe we should the LV to the size the user asked for?

@cfergeau then what if user want to extend the disk size because there is no free space on VG for persistent volume for the workload? Right now I am putting half for each so that user doesn't need to worry about in both the cases otherwise the correct solution would be add config option which user provide if he wants LV to increase or just want free size for it's persistent volume workload.

cfergeau commented:
> > Maybe we should the LV to the size the user asked for?
> 
> @cfergeau then what if user want to extend the disk size because there is no free space on VG for persistent volume for the workload? Right now I am putting half for each so that user doesn't need to worry about in both the cases otherwise the correct solution would be add config option which user provide if he wants LV to increase or just want free size for it's persistent volume workload.

This is not answering my initial question. If I run `crc config set disk-size xx` what would users expect to happen? Once we've answered this question, we can discuss some more the implications with the free space needed by the CSI driver.

praveenkumar commented:
/hold

we need to understand what user scenario we want to implement https://github.com/crc-org/crc/issues/3624#issuecomment-1560892999

praveenkumar committed:
machine: Refactor disk resize code

Add `getrootPartition` and `runGrowpart` helper which is used in
follow-up commit.

praveenkumar committed:
Microshift: add disk resize functionality

The way this PR works is in following steps
- It grows the disk size using growpart command
- Resize the physical volume using pvresize command
- Get the size of whole physical volume and root logical volume then we
  subtract existing free volume and check how much lv can be extended.

Note: This is not going to extend the free space which is used by the
LVM Storage Operator for persistent local storage

praveenkumar commented:
/unhold

implemented as per https://github.com/crc-org/crc/issues/3624#issuecomment-1562708790 one.

gbraad commented on pkg/crc/machine/start.go:
```diff
@@ -134,19 +158,55 @@ func growRootFileSystem(sshRunner *crcssh.Runner, preset crcPreset.Preset) error
 	}
 
+	// vgFree space as part of the bundle is default to ~15.02G (16127098880 byte)
+	vgFree := 16127098880
```
magic value
